File: photoeditor/src/main/java/ja/burhanrashid52/photoeditor/shape/AbstractShape.java
package ja.burhanrashid52.photoeditor.shape;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RectF;

abstract public class AbstractShape implements Shape {

    protected float TOUCH_TOLERANCE = 4;

    protected Path path = new Path();
    protected float left, top, right, bottom;
    protected abstract String getTag();

    @Override
    public void draw(Canvas canvas, Paint paint) {
        canvas.drawPath(path, paint);
    }

    public RectF getBounds() {
        RectF bounds = new RectF();
        path.computeBounds(bounds, true);
        return bounds;
    }

    public boolean hasBeenTapped() {
        RectF bounds = getBounds();
        return bounds.top < TOUCH_TOLERANCE &&
                bounds.bottom < TOUCH_TOLERANCE &&
                bounds.left < TOUCH_TOLERANCE &&
                bounds.right < TOUCH_TOLERANCE;
    }

    public String toString() {
        return getTag() +
                ": left: " + left +
                " - top: " + top +
                " - right: " + right +
                " - bottom: " + bottom;
    }

}

File: photoeditor/src/main/java/ja/burhanrashid52/photoeditor/shape/AbstractShape.kt
package ja.burhanrashid52.photoeditor.shape

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF

abstract class AbstractShape(protected val tag: String) : Shape {
    protected var path = Path()
    protected var left = 0f
    protected var top = 0f
    protected var right = 0f
    protected var bottom = 0f

    override fun draw(canvas: Canvas, paint: Paint) {
        canvas.drawPath(path, paint)
    }

    private val bounds: RectF
        get() {
            val bounds = RectF()
            path.computeBounds(bounds, true)
            return bounds
        }

    fun hasBeenTapped(): Boolean {
        val bounds = bounds
        return bounds.top < TOUCH_TOLERANCE && bounds.bottom < TOUCH_TOLERANCE && bounds.left < TOUCH_TOLERANCE && bounds.right < TOUCH_TOLERANCE
    }

    override fun toString(): String {
        return tag +
                ": left: " + left +
                " - top: " + top +
                " - right: " + right +
                " - bottom: " + bottom
    }

    companion object {
        const val TOUCH_TOLERANCE = 4f
    }
}